elasticsearch/esql/esql.py

```diff
@@ -1046,6 +1095,65 @@ def _render_internal(self) -> str:
         return f'RENAME {", ".join([f"{self._format_id(old_name)} AS {self._format_id(new_name)}" for old_name, new_name in self._columns.items()])}'
 
 
+class Rerank(ESQLBase):
+    """Implementation of the ``RERANK`` processing command.
+
+    This class inherits from :class:`ESQLBase <elasticsearch.esql.esql.ESQLBase>`,
+    to make it possible to chain all the commands that belong to an ES|QL query
+    in a single expression.
+    """
+
+    def __init__(
+        self, parent: ESQLBase, *query: ExpressionType, **named_query: ExpressionType
+    ):
+        if len(query) + len(named_query) > 1:
+            raise ValueError(
+                "this method requires either one positional or one keyword argument only"
+            )
+        super().__init__(parent)
+        self._query = query
+        self._named_query = named_query
+        self._fields: Optional[Tuple[str, ...]] = None
+        self._inference_id: Optional[str] = None
+
+    def on(self, *fields: str) -> "Rerank":
+        self._fields = fields
+        return self
+
+    def with_(self, inference_id: str) -> "Rerank":
+        """Continuation of the `COMPLETION` command.
+
+        :param inference_id: The ID of the inference endpoint to use for the task. The
+                             inference endpoint must be configured with the completion
+                             task type.
+        """
+        self._inference_id = inference_id
+        return self
+
+    def _render_internal(self) -> str:
+        if self._fields is None:
+            raise ValueError(
+                "The rerank command requires one or more fields to rerank on"
+            )
+        if self._inference_id is None:
+            raise ValueError("The completion command requires an inference ID")
+        with_ = {"inference_id": self._inference_id}
+        if self._named_query:
+            column = list(self._named_query.keys())[0]
+            query = list(self._named_query.values())[0]
+            return (
+                f"RERANK {self._format_id(column)} = {json.dumps(query)} "
+                f"ON {', '.join([self._format_id(field) for field in self._fields])} "
+                f"WITH {json.dumps(with_)}"
+            )
+        else:
+            return (
+                f"RERANK {json.dumps(self._query[0])} "
+                f"ON {', '.join([self._format_id(field) for field in self._fields])} "
+                f"WITH {json.dumps(with_)}"
```

Comment:
nit: the code duplication makes me slightly nervous, given this code is not generated

Reply:
I thought it best to keep the templates for each version of the command separate so that you can see what the command looks like more easily, but sure, it does add some repetition. I'll see if I can come up with a better way to structure this and other similar commands.